2853/duplicates pages

dominikx96 opened the pull request:
# Pull Request Template

## Issue Overview

This PR addresses #2853

- [x] This change addresses the issue in full
- [ ] This change addresses only certain aspects of the issue
- [ ] This change is a dependency for another issue
- [ ] This change has a dependency from another issue

## Description

Creates new views for all, pending and resolved applications. 

## How Can This Be Tested/Reviewed?

1. Open the partners portal and choose a listing
2. Click on the "open" cell
3. The URL should be: `/listings/{listing_id}/applications` and "Applications" should be an active element in breadcrumbs
4. Open the `/listings/{listing_id}/applications/pending`, and make sure "Pending" is an active element in breadcrumbs
5. Open the `/listings/{listing_id}/applications/resolved`, and make sure "Resolved" is an active element in breadcrumbs

Describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have added QA notes to the issue with applicable URLs
- [x] I have performed a self-review of my own code
- [x] I have reviewed the changes in a desktop view
- [x] I have reviewed the changes in a mobile view
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] I have assigned reviewers
- [ ] I have run `yarn generate:client` and/or created a migration if I made backend changes that require them
- [ ] I have exported any new pieces added to ui-components
- [x] My commit message(s) is/are polished, and any breaking changes are indicated in the message and are well-described
- [x] Commits made across packages purposefully have the same commit message/version change, else are separated into different commits

## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue, and ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.

dominikx96 commented:
We've merged multi-level nav so I'm going to update that and ask for a review after.

dominikx96 commented:
@YazeedLoonat frontend part is ready to integrate, feel free to add a backend part :)

emilyjablonski commented on backend/core/src/seeder/seeds/applications.ts:
```diff
@@ -226,15 +229,70 @@ export const makeNewApplication = async (
     confirmationCode: ApplicationsService.generateConfirmationCode(),
   })
+
+  if (pos === 0) {
```
I thought we were changing all of the duplicate finding logic and that that work was TBD? Just want to make sure we aren't doing work we won't need in general for the backend.

seanmalbert replied:
I think this should be ok, since it's manually saving to the afs repo. The new logic, with the first iteration anyway, mostly just moves the logic to a cron job with some performance improvements. We will have to make an update, since it adds a new key, but it should be a quick update in either PR, depending on which one gets merged first.

emilyjablonski commented on ui-components/src/page_components/NavigationHeader.tsx:
```diff
@@ -42,11 +43,19 @@ const NavigationHeader = ({
       {
         label: tabs?.applicationsLabel || "",
         path: `/listings/${listingId}/applications`,
+        activePaths: [
```
Definitely a problem for different PR, but eventually we'll need to strip these implementation details out of ui-components

seanmalbert replied:
https://github.com/bloom-housing/bloom/issues/2962

emilyjablonski commented on sites/partners/pages/listings/[id]/applications/pending/index.tsx:
```diff
@@ -0,0 +1,187 @@
+import React from "react"
```
This file and the resolved file are almost identical, can we extract the shared page content into a shared component in partners?

emilyjablonski reviewed (changes requested):
Some questions!

*  The issue says to not include subcategories for rule, but I see them here (and I do also see them in the Figma, so there is a discrepancy) - @kathyccheng What is preferred?
* If we do decide to show those subcategories, it looks like they should not be shown until the parent is selected, and once one is selected we should reset the focus state so it doesn't say on the nav link
* On a reseed for Coliseum, pending review lists 4 apps, but nothing shows, is that expected?
* `All Applications` should be capitalized

ludtkemorgan commented on backend/core/src/application-flagged-sets/application-flagged-sets.service.ts:
```diff
@@ -364,4 +365,60 @@ export class ApplicationFlaggedSetsService {
+    const pendingNameQB = this.afsRepository.createQueryBuilder("afs")
+    pendingNameQB.select("SUM(1) as totalNamePendingCount")
+    pendingNameQB.where("afs.listingId = :listingId", { listingId: queryParams.listingId })
+    pendingNameQB.andWhere("afs.status = :status", { status: FlaggedSetStatus.flagged })
+    pendingNameQB.andWhere("afs.rule = :rule", { rule: Rule.nameAndDOB })
+
+    const pendingEmailQB = this.afsRepository.createQueryBuilder("afs")
+    pendingEmailQB.select("SUM(1) as totalEmailPendingCount")
+    pendingEmailQB.where("afs.listingId = :listingId", { listingId: queryParams.listingId })
+    pendingEmailQB.andWhere("afs.status = :status", { status: FlaggedSetStatus.flagged })
+    pendingEmailQB.andWhere("afs.rule = :rule", { rule: Rule.email })
+
+    const results = await Promise.all(
+      [allQB, resolvedQB, pendingQB, pendingNameQB, pendingEmailQB].map(
+        async (query) => await query.getRawOne()
+      )
+    )
+
+    const res: ApplicationFlaggedSetMeta = {}
+
+    results.forEach((elem) => {
+      if ("totalcount" in elem) {
+        res.totalCount = elem.totalcount
+      }
+      if ("totalresolvedcount" in elem) {
+        res.totalResolvedCount = elem.totalresolvedcount
+      }
+      if ("totalpendingcount" in elem) {
+        res.totalPendingCount = elem.totalpendingcount
+      }
+      if ("totalnamependingcount" in elem) {
+        res.totalNamePendingCount = elem.totalnamependingcount
+      }
+      if ("totalemailpendingcount" in elem) {
+        res.totalEmailPendingCount = elem.totalemailpendingcount
+      }
+    })
+
+    return res
+  }
```
There is a lot of duplicated code here, mostly around similar db queries and getting the result values out. Could it potentially be cleaned up to something like this?
```suggestion
    async meta(queryParams: PaginatedApplicationFlaggedSetQueryParams) {
    const constructQuery = (params: { listingId: string, status?: FlaggedSetStatus, rule?: Rule }): SelectQueryBuilder<Application> => {
      const query = this.applicationsRepository.createQueryBuilder("afs")
      query.select("SUM(1) as value").where("afs.listing_id = :listingId", { listingId: params.listingId })
      if (params.status) {
        query.andWhere("afs.status = :status", { status: params.status})
      }
      if (params.rule) {
        console.log(params.rule)
        query.andWhere("afs.rule = :rule", { rule: params.rule })
      }
      return query
    }

    const allQB = this.applicationsRepository.createQueryBuilder("afs")
    allQB.select("SUM(1) as value")
    allQB.where("afs.listing_id = :listingId", { listingId: queryParams.listingId })
    const resolvedQB = constructQuery({listingId: queryParams.listingId, status: FlaggedSetStatus.resolved})
    const pendingQB = constructQuery({listingId: queryParams.listingId, status: FlaggedSetStatus.flagged})
    const pendingNameQB = constructQuery({listingId: queryParams.listingId, status: FlaggedSetStatus.flagged, rule: Rule.nameAndDOB})
    const pendingEmailQB = constructQuery({listingId: queryParams.listingId, status: FlaggedSetStatus.flagged, rule: Rule.email})


    const [totalCount, totalResolvedCount, totalPendingCount, totalNamePendingCount, totalEmailPendingCount] = await Promise.all(
      [allQB, resolvedQB, pendingQB, pendingNameQB, pendingEmailQB].map(
        async (query) => await query.getRawOne()
      )
    )

    const res: ApplicationFlaggedSetMeta = {
      totalCount: totalCount.value,
      totalResolvedCount: totalResolvedCount.value,
      totalPendingCount: totalPendingCount.value,
      totalNamePendingCount: totalNamePendingCount.value,
      totalEmailPendingCount: totalEmailPendingCount.value
    }

    return res
  }
```

seanmalbert replied:
@YazeedLoonat ^

YazeedLoonat commented:
Hey @ludtkemorgan nice call, got this updated, lemme know what you think

ludtkemorgan commented:
> Hey @ludtkemorgan nice call, got this updated, lemme know what you think

Thanks for making those changes. It looks good

emilyjablonski commented:
@YazeedLoonat @dominikx96 Just a reminder there are still two questions / requested changes I have above to DRY out the new page and capitalize the string!

seanmalbert commented:
Answers for @emilyjablonski and things to do for @dominikx96 ,
> * The issue says to not include subcategories for rule, but I see them here (and I do also see them in the Figma, so there is a discrepancy) - @kathyccheng What is preferred?
> * If we do decide to show those subcategories, it looks like they should not be shown until the parent is selected, and once one is selected we should reset the focus state so it doesn't stay on the nav link

Kathy says to hide these for now.

> * On a reseed for Coliseum, pending review lists 4 apps, but nothing shows, is that expected?

This is expected, as they're handled in a separate PR.

> * `All Applications` should be capitalized

👍

dominikx96 commented:
@emilyjablonski  I updated the "All Applications" translation.

Also, refactored pages `pending` and `resolved` - I moved table columns and link formatter parts to a separate file. I think we should keep the page structure as it is, it's not complex and it gives us more flexibility. I'm afraid of moving both pages to separate layouts, the current structure vs. a new layout with a callback might only complicate the code. I am open to discussion :)

Optionally we can create a new component to replace an `<article>` tag because tailwind classes are the same for 3 pages.

@YazeedLoonat I connected these pages to `useFlaggedApplicationsList` hook, but I noticed the API doesn't return data in proper format:

![Screenshot 2022-08-17 at 10 58 29](https://user-images.githubusercontent.com/16655908/185084216-28a1dc1d-482c-485d-86bc-c337181fccfd.png)

We need to somehow fill columns:
- duplicate group
- primary applicant
- pending review

and support (see screenshot below):
- last updated
- trigger duplicates scan 

![Screenshot 2022-08-17 at 11 28 05](https://user-images.githubusercontent.com/16655908/185084878-ad950a21-4dfa-410c-8940-dd8ccf50cd82.png)

and add a filter to fetch apps by its rule (email/name + dob). From the frontend perspective, we have the following parameters to determine which type of apps we should fetch:
- /applications/pending?type=name_dob
- /applications/pending?type=email

I imagine we might provide this query param along with an API call.

YazeedLoonat commented:
Hey @dominikx96 a lot of the work for the backend calls was set up: https://github.com/bloom-housing/bloom/pull/2944 I added the requests to the backend as well as the backend code that would get that data properly 

we could probably pull some of that to here, but wasn't sure if that should be a part of this pr since the AC seemed pretty specific about the side nav and not so much the data in the columns


lemme know if it makes sense to ya to keep those backend updates in the follow up pr or if you think we should move it into here

seanmalbert commented:
@YazeedLoonat and @dominikx96 , for last updated and trigger duplicates scan, that can be handled in #2961. Once the rest of this is approved, we can merge that into this branch and hook them up before merging this.

emilyjablonski commented:
@dominikx96 Per Sean's comment we need to hide the subcategories for rules, otherwise lgtm!
